### PR TITLE
CV Batch Footage Extraction

### LINKED ISSUE #407: Batch Footage Extraction
# Background

CV collects footage in the form of bag files and video files. Then, frames need to be extracted from the footage in the form of individual images, which are then uploaded to Roboflow and labeled. Currently, we have no unified script to extract frames from bag and video files and our scripts are limited to only extracting from one file at a time. Since we expect to collect a large amount of footage in both formats, we need a script that can extract frames from multiple files of different types.

Additionally, we want to avoid labeling images that are very similar to each other, as they provide redundant information to our model and may lead to overfitting. Currently, we have no automated solution for this. In the past, for example, if we had recorded footage at 30fps, we would save every 10th frame and discard the rest since the time difference between frames is so small. While this simple approach filters out many repetitive frames, some still fall through the cracks. For example, if the footage has a long period with no movement, then for that period even every 10th frame would be similar. Instead of simply skipping frames, we can use a smarter approach – compute the similarity between images and discard ones that are similar.

# Purpose

Develop a script that extracts frames from multiple files of different types and filters out similar frames.

### PR DESCRIPTION
**Pull request by vedarshshah**
CV developed the script `extract_footage.py` to enable us to take video and bag files in batch, extract individual images from this files, and upload those frames to our dataset on Roboflow for labeling. Closes #407.

### CODE REVIEW

**Review by Brokemia — COMMENTED** [review]
Just a few small comments

File: onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py [hunk 1]
```diff
@@ -1,78 +1,459 @@
...
         :param step_size: Will extract a frame every step_size number of frames.
-        :return: Will
         """
 
-        file_basename = os.path.basename(footage_path)
-        file_name_no_extension = os.path.splitext(file_basename)[0]
-
         bag = rosbag.Bag(footage_path)
 
+        num_frames = bag.get_message_count('/'+topic_name)
```
[the remark this answers is not in the document]
Comment (maxwellmlin) [reply onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py:41]: Slash has been removed.

File: onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py [hunk 2]
```diff
@@ -1,78 +1,459 @@
...
+
+            if success:
+                status['successful'].append(image)
+                shutil.move(image_path, uploaded_dir)
+            else:
+                status['unsuccessful'].append(image)
+
+        return status
+
+    def upload_images_to_roboflow_with_success(self, rf_project, image_path, batch_name):
```
[the remark this answers is not in the document]
Comment (maxwellmlin) [reply onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py:140]: Renamed.

File: onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py [hunk 3]
```diff
@@ -1,78 +1,459 @@
...
+                    continue
+
+                # extracting file_name
+                file_name = file[::-1].replace('_', '.', 1)[::-1]
+                footage_path = os.path.join(directory, file_name)
+                extracted_footage_path = os.path.join(EXTRACTED_FILES_DIR, file_name)
+
+                if file_name.endswith('bag'):
+                    for topic in file_dict:
+                        if topic in ['enabled', 'use_name']:  # The keys 'enabled' and 'use_name' are not topics
```
[the remark this answers is not in the document]
Comment (maxwellmlin) [reply onboard/catkin_ws/src/cv/footage_extraction/extract_footage.py]: Done. `create_footage_extraction_config_file` nows creates a `topics` field which `extract_footage_with_config` reads from.

**Comment by maxwellmlin** [comment]
@Brokemia All review comments have been addressed. Please take a look.

**Review by Brokemia — APPROVED** [review]
Looks good

**Pull request merged**